lib/Sema/ConstraintSystem.cpp

```diff
@@ -1015,8 +1015,42 @@ static void bindArchetypesFromContext(
       // parameters, or because this generic parameter was constrained
       // away into a concrete type.
       if (found != replacements.end()) {
+        Type contextTy;
+
+        if (genericEnv) {
+          contextTy = genericEnv->mapTypeIntoContext(paramTy);
+        } else {
+          auto &TC = cs.TC;
```

Comment:
The general approach looks good, but please pull this computation out into a separate function.

Reply:
I'm going to make `getInnermostConformingDC` which returns `NominalTypeDecl *`.